src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -362,32 +362,35 @@ class AuthenticationCompleteException : Exception { }
 
         string? GetPassword(string username)
         {
-            var passwd = Settings.Password;
-            if (passwd != null)
-                return passwd;
+            var password = Settings.Password;
+            if (password != null)
+                return password;
 
             // No password was provided. Attempt to pull the password from the pgpass file.
-            var matchingEntry = PgPassFile.Load(Settings.Passfile)?.GetFirstMatchingEntry(Host, Port, Settings.Database!, username);
-            if (matchingEntry != null)
+            var passFilePath = Settings.Passfile ?? PostgresEnvironment.PassFile ?? PostgresEnvironment.PassFileDefault;
+            if (passFilePath != null && File.Exists(passFilePath))
             {
-                Log.Trace("Taking password from pgpass file");
-                return matchingEntry.Password;
-            }
-
-            if (ProvidePasswordCallback != null)
-            {
-                Log.Trace($"Taking password from {nameof(ProvidePasswordCallback)} delegate");
-                try
-                {
-                    return ProvidePasswordCallback(Host, Port, Settings.Database!, username);
-                }
-                catch (Exception e)
+                var matchingEntry = new PgPassFile(passFilePath)
+                    .GetFirstMatchingEntry(Host, Port, Settings.Database!, username);
+                if (matchingEntry != null)
                 {
-                    throw new NpgsqlException($"Obtaining password using {nameof(NpgsqlConnection)}.{nameof(ProvidePasswordCallback)} delegate failed", e);
+                    Log.Trace("Taking password from pgpass file");
+                    return matchingEntry.Password;
                 }
             }
 
-            return null;
+            if (ProvidePasswordCallback is null)
+                return PostgresEnvironment.Password;
+
+            Log.Trace($"Taking password from {nameof(ProvidePasswordCallback)} delegate");
```

Comment:
Didn't we say that the password callback should take precedence over environment-provided pgpass?

Basically I think the precedence should look like this, but we should probably see what the libpq behavior is (where relevant):

1. Password in connection string
2. Passfile in connection string
3. Password callback (not so sure between this and 2 above)
4. Password in environment variable
5. Passfile in environment variable
6. Default passfile

Reply:
Won't do it here because this pull requests adds two new variables only and the requested change is out of scope. Tracked by #2695.